### DISCUSSION TITLE
File-based routing (without vite)

### DISCUSSION THREAD

**Discussion by kiki-kanri** [honojs/hono#3508] [General]
I'm writing a backend application and I'd like to use file-based routing, similar to how nitrojs uses pathname.method.ts to define routes and auto-scan for registrations.

So far I've found a way to use honox, but it's based on vite to implement file routing, and as a pure backend I don't want to use extra features like tsx.

Is there another way to use file routing, or do I have to implement it myself?

**Comment by maou-shonen** [comment]
If you are using Bun, there's a [FileSystemRouter](https://bun.sh/docs/api/file-system-router) available.
If you're okay with having an unused index.html file, Vite + Hono + vanilla JS is actually a good option.

**Reply by kiki-kanri** [reply]
Thanks for your reply and letting me know that FileSystemRouter is available.

But my goal is to implement it without vite, simply using only bun+hono, etc. Someone down there has the preliminary to make this package.

However, at this stage it is estimated that a number of core functions will have to be implemented in order to make it stable and production-ready.

I will try to use other methods to solve or realize it, thanks again for your reply!

**Comment by thisjt** [comment]
I'm currently developing a file-based router plugin for hono.

https://www.npmjs.com/package/hono-file-router

It's not yet production-ready as I lit just made this a few days ago as I wanted a file-based router. Let me know what you think.

**Reply by kiki-kanri** [reply]
I'm also currently using the custom scan file and register routing feature to realize file-based routing, see [this repo](https://github.com/kiki-kanri/auto-hono).

However, due to file-based reasons, there are a few problems:

- Before using bun build, you need to write additional file scans and replace the existing routing registry so that bun can read all the files and package them.
- The original type of function is missing.
-  Some of the TODOs mentioned on your plugin.

Plus since the repository is the base for my project development, I often use packages that require workers after I fork it out, which bun doesn't support yet, so I'm going to migrate to the hyper-express framework again, but still use file routing.

My idea is to integrate zod-validator and zod-openapi functionality to auto-route registrations and switch hono servers as needed to use the corresponding functionality.

There is also the possibility of adding a custom attribute to the final route handler and then accessing that attribute in the middleware for identity and permission determination.

But I guess I'll have more ideas when I rewrite it with hyper-express, but if your package is then stable and the core issues are resolved, I should use it on other small, bun-executable projects.
